### DISCUSSION TITLE
Strongly typed IDs

### DISCUSSION THREAD

**Discussion by DirtyNative** [reactiveui/refit#1527] [Q&A] [answered]
I am wondering how to get strongly typed IDs to work with Refit. I am using [this package](https://www.nuget.org/packages/Meziantou.Framework.StronglyTypedId) to automatically generate all needed things for the IDs. Here are the classes:

```
[StronglyTypedId(typeof(Guid))]
public partial struct UserId
{
	public partial class UserIdTypeConverter { }
	public partial class UserIdJsonConverter { }
	public partial class UserIdNewtonsoftJsonConverter { }
}
```

```
public interface IAudioSourceRefitClient
{
	[Get($"{Routes.AudioSourceRoutes.BaseRoute}/{Routes.AudioSourceRoutes.GetAllByUserRoute}")]
	Task<IEnumerable<AudioSource>> GetAllByUserId(UserId userId, CancellationToken cancellationToken);
}
```

This works quiet fine as long as I define the "userId" as a Guid. Preferably when working with strongly typed IDs I want to use the UserId here, but this results in Exceptions which I think Refit is producing.

`
Refit.ApiException: Response status code does not indicate success: 400 (Bad Request).
         at Refit.RequestBuilderImplementation.<>c__DisplayClass14_0`2.<<BuildCancellableTaskFuncForMethod>b__0>d.MoveNext() in /_/Refit/RequestBuilderImplementation.cs:line 288
      --- End of stack trace from previous location ---
`

I guess Refit is not able to convert the UserId correctly to a Guid, even that I have included the Converters within my RefitSettings. Do I miss something here or is Refit not meant to be used with this feature?

**Comment by snailcatcher** [comment]
@DirtyNative Did you find a solution for your problem?
I'm struggling with exactly the same issue.

I have one addition:
Refit will serializes the StronglyTypedId to something like `{Value: "bla", ValueAsString: "bla"}`.
Actually it is expected to only get `bla` or at least the result of the `.ToString()` method (which can be modified by the `GenerateToStringAsRecord` setting for the StronglyTypedId)

**Comment by glennawatson** [comment, marked as answer]
The converters you set in `RefitSettings` (JsonConverter / TypeConverter) only affect JSON body serialization and TypeConverter-based binding. They do not control how a value is rendered into a URL path or query string.

For URL path/query parameters Refit uses `IUrlParameterFormatter`. The `DefaultUrlParameterFormatter` formats with `string.Format("{0}", value)`, i.e. it calls the value's `ToString()`. So:

- If your `UserId` has a `ToString()` that returns the raw guid, the path will be correct.
- The reason snailcatcher sees `{Value: "...", ValueAsString: "..."}` is that the struct is being object-serialized (its default ToString, or a query-object being JSON-shaped) instead of producing the bare value.

Fixes:

1. Make the strongly typed ID render correctly in URLs by ensuring `ToString()` returns just the value. With Meziantou.Framework.StronglyTypedId you can control this (e.g. it generates a ToString that returns the inner value; verify your `GenerateToStringAsRecord` setting is not producing the record-style `{ Value = ... }` output).

2. Or supply a custom `IUrlParameterFormatter` in `RefitSettings.UrlParameterFormatter` that detects your ID types and emits the inner value:

```csharp
public class StronglyTypedIdUrlFormatter : DefaultUrlParameterFormatter
{
    public override string? Format(object? value, ICustomAttributeProvider p, Type type)
    {
        if (value is UserId id)
            return id.Value.ToString();
        return base.Format(value, p, type);
    }
}
```

For the 400 in the original post: the server received something other than a bare guid in the path, which is the same root cause. Once the URL formatting emits the raw value the request will match.
